apollo-runtime-kotlin/src/iosMain/kotlin/com/apollographql/apollo/network/ApolloHttpNetworkTransport.kt

```diff
@@ -0,0 +1,208 @@
+    }
+  }
+
+  private fun GraphQLRequest.toHttpGetRequest(): NSMutableURLRequest {
+    val urlComponents = NSURLComponents(uRL = serverUrl, resolvingAgainstBaseURL = false)
+    urlComponents.queryItems = listOfNotNull(
+        NSURLQueryItem(name = "query", value = document),
+        NSURLQueryItem(name = "operationName", value = operationName),
+        if (variables.isNotEmpty()) NSURLQueryItem(name = "variables", value = variables) else null
+    )
+    return NSMutableURLRequest.requestWithURL(urlComponents.URL!!).apply {
+      setHTTPMethod("GET")
+      httpHeaders.forEach { (key, value) -> setValue(value, forHTTPHeaderField = key) }
+      setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+    }
+  }
+
+  private fun GraphQLRequest.toHttpPostRequest(): NSMutableURLRequest {
+    return NSMutableURLRequest.requestWithURL(serverUrl).apply {
+      val buffer = Buffer()
+      JsonWriter.of(buffer)
+          .beginObject()
+          .name("operationName").value(operationName)
+          .name("query").value(document)
+          .name("variables").value(variables)
+          .endObject()
+          .close()
+      val postBody = buffer.readByteArray().toNSData()
+
+      setHTTPMethod("POST")
+      httpHeaders.forEach { (key, value) -> setValue(value, forHTTPHeaderField = key) }
+      setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+      setHTTPBody(postBody)
+    }
+  }
+
+  private fun parse(
+      data: NSData?,
+      httpResponse: NSHTTPURLResponse,
+      error: NSError?
+  ): Result {
+    if (error != null) return Result.Failure(
+        ApolloException(
+            message = "Failed to execute GraphQL http network request",
+            error = ApolloError.Network,
+            cause = IOException(error.localizedDescription)
+        )
+    )
+
+    val statusCode = httpResponse.statusCode.toInt()
+    if (statusCode !in 200..299) return Result.Failure(
+        ApolloException(
+            message = "Http request failed with status code `$statusCode`",
+            error = ApolloError.Network
+        )
+    )
+
+    if (data == null) return Result.Failure(
+        ApolloException(
+            message = "Failed to parse GraphQL http network response: EOF",
+            error = ApolloError.Network
+        )
+    )
+
+    return Result.Success(data)
+  }
+
+  sealed class Result {
+    class Success(val data: NSData) : Result()
+    class Failure(val cause: ApolloException) : Result()
+  }
+
+  private fun ByteArray.toNSData(): NSData = NSMutableData().apply {
+    if (isEmpty()) return@apply
+    this@toNSData.usePinned {
+      appendBytes(it.addressOf(0), size.convert())
+    }
+  }
+
+  @Suppress("NAME_SHADOWING")
+  private fun Result.dispatchOnMain(producerRef: COpaquePointer) {
+    val producerWithResultRef = StableRef.create((producerRef to this).freeze())
+    dispatch_async_f(
+        queue = dispatch_get_main_queue(),
+        context = producerWithResultRef.asCPointer(),
```

Comment:
`context` and `queue` here are properties of `NetworkTransport`, right? I *think* you're OK in terms of automatic reference counting here, but it might be worth doing some profiling to make sure you're not creating retain cycles.

Reply:
>context and queue here are properties of NetworkTransport, right

not sure I follow, `queue` and `context ` are named arguments for `dispatch_async_f ` C function. 
`dispatch_get_main_queue()` - gives access to main queue
`producerWithResultRef.asCPointer()` - just a pointer

Reply:
ah ok-  I wasn't quite clear on the fact that they were named arguments for the function rather than vars declared elsewhere. Thanks!